src/System.Private.CoreLib/shared/System/Number.Parsing.cs

```diff
@@ -1468,64 +1468,87 @@ internal static decimal ParseDecimal(ReadOnlySpan<char> value, NumberStyles styl
 
         private static unsafe bool NumberBufferToDecimal(ref NumberBuffer number, ref decimal value)
         {
-            decimal d = new decimal();
-
             char* p = number.digits;
             int e = number.scale;
-            if (*p == 0)
+            bool sign = number.sign;
+            uint c = *p;
+            if (c == 0)
             {
                 // To avoid risking an app-compat issue with pre 4.5 (where some app was illegally using Reflection to examine the internal scale bits), we'll only force
                 // the scale to 0 if the scale was previously positive (previously, such cases were unparsable to a bug.)
-                if (e > 0)
+                value = new decimal(0, 0, 0, sign, (byte)Math.Clamp(-e, 0, 28));
+                return true;
+            }
+
+            if (e > DecimalPrecision)
+                return false;
+
+            ulong low64 = 0;
+            while (e > -28)
+            {
+                e--;
+                low64 *= 10;
+                low64 += c - '0';
+                c = *++p;
+                if (low64 >= ulong.MaxValue / 10)
+                    break;
+                if (c == 0)
                 {
-                    e = 0;
+                    while (e > 0)
+                    {
+                        e--;
+                        low64 *= 10;
+                        if (low64 >= ulong.MaxValue / 10)
+                            break;
+                    }
+                    break;
                 }
             }
-            else
+
+            uint high = 0;
+            while ((e > 0 || c != 0 && e > -28) &&
+              (high < uint.MaxValue / 10 || high == uint.MaxValue / 10 && (low64 < 0x99999999_99999999 || low64 == 0x99999999_99999999 && c <= '5')))
             {
-                if (e > DecimalPrecision)
-                    return false;
+                // multiply by 10
+                ulong tmpLow = (uint)low64 * 10UL;
+                ulong tmp64 = (uint)(low64 >> 32) * 10UL + (tmpLow >> 32);
+                low64 = (uint)tmpLow + (tmp64 << 32);
+                high = (uint)(tmp64 >> 32) + high * 10;
 
-                while (((e > 0) || ((*p != 0) && (e > -28))) &&
-                       ((d.High < 0x19999999) || ((d.High == 0x19999999) &&
-                                                  ((d.Mid < 0x99999999) || ((d.Mid == 0x99999999) &&
-                                                                            ((d.Low < 0x99999999) || ((d.Low == 0x99999999) &&
-                                                                                                      (*p <= '5'))))))))
+                if (c != 0)
                 {
-                    decimal.DecMul10(ref d);
-                    if (*p != 0)
-                        decimal.DecAddInt32(ref d, (uint)(*p++ - '0'));
-                    e--;
+                    c -= '0';
+                    if ((low64 += c) < c)
```

Comment:
Is inlining `low64 += c` a win here? It looks like trying to hide something to me. Why not:
```
low64 += c;
if (low64 < c)
```

Reply:
Assignments in conditional statements are terrible and should be avoided.